src/psyclone/psyir/frontend/fparser2.py

```diff
@@ -3274,21 +3278,274 @@ def _if_stmt_handler(self, node, parent):
+        # capturing the name of the type or class clauses (created in
+        # step2).
+        code = "program dummy\n"
+        code += f"select type({selector})\n"
+        for idx in range(select_idx+1):
+            if idx == default_idx:
+                pointer_symbols.append(None)
+                continue
+            pointer_name = parent.scope.symbol_table.next_available_name(
+                f"ptr_{guard_type_repr[idx]}")
+            tmp = f"{guard_type[idx]}"
+            if not intrinsic_type_name[idx]:
+                tmp = f"type({tmp})"
+            pointer_type = UnknownFortranType(
+                f"{tmp}, pointer :: {pointer_name}\n")
+            if guard_type[idx] in ["CHARACTER(LEN = *)", "CHARACTER*(*)"]:
+                pointer_symbol = DataSymbol(
+                    pointer_name, pointer_type,
+                    interface=ArgumentInterface(
+                        ArgumentInterface.Access.WRITE))
+            else:
+                pointer_symbol = DataSymbol(pointer_name, pointer_type)
+
+            parent.scope.symbol_table.add(pointer_symbol)
+            if guard_type[idx] in ["CHARACTER(LEN = *)", "CHARACTER*(*)"]:
+                # pylint: disable=protected-access
+                parent.scope.symbol_table._argument_list.append(pointer_symbol)
+            pointer_symbols.append(pointer_symbol)
+            if (intrinsic_type_name[idx] and
+                    intrinsic_type_name[idx].lower() == "character"):
+                # The type spec for a character intrinsic must always
+                # be assumed
+                type_spec = "CHARACTER(LEN = *)"
+            else:
+                # Use the variable declaration
+                type_spec = guard_type[idx]
+            code += f"  {clause_type[idx]} ({type_spec})\n"
+            code += (f"    {type_string_name} = "
+                     f"\"{guard_type_repr[idx].lower()}\"\n")
+            code += (f"    {pointer_name} => {selector}\n")
+        code += "end select\n"
+        code += "end program\n"
+        parser = ParserFactory().create(std="f2008")
+        reader = FortranStringReader(code)
+        fp2_program = parser(reader)
+        # Only use the select type clause within the fparser2 program
+        fp2_select_type = fp2_program.children[0].children[1]
+        code_block = CodeBlock(
+            [fp2_select_type], CodeBlock.Structure.STATEMENT, parent=parent)
+        # Handlers assume a single node is returned and in this
+        # implementation we create an assignment, a CodeBlock and an
+        # if statement (see later). Therefore we add the assignment
+        # and codeblock to the parent here and return the if statement
+        # in the usual way.
+        parent.addchild(Assignment.create(
+            Reference(type_string_symbol), Literal("", CHARACTER_TYPE)))
+        parent.addchild(code_block)
+
+        # Step4: Create the (potentially nested) if statement that
```

Comment:
Could this step be broken out as a separate utility that is just passed the list(s) describing the select-type branches?

Reply:
I've split all the stages up. Not sure whether this is better or worse so I'll leave it to the reviewer to decide.